### PR TITLE
Make string-format unit tests more readable

### LINKED ISSUE #145: Make unit tests easier to understand
I was about to dig into an open issue when I found that I was having difficulty reading reformat-string unit tests, hence this proposal. 

Inspired by the conventions in #88, I think we can make reformat-string unit tests easier to digest by:
1. Giving each line its own line (no longer embedding \n)
1. Making the intent of each test clear by giving it a description. Having tests separated into testing block helped, but sometimes I wondered what an individual test was trying to achieve.
1. Making failures easier to decipher. When a reformat-string test failed, I was left confused between actual and expected results; they seem reversed. Note that I am using [eftest](https://github.com/weavejester/eftest) for friendlier test output.

I'm using this issue to propose the change. It seems like a good idea to me, what do you think? If you like it, I'm happy to do the work.

# Simple test example
## Code
Current
```clojure
    (is (= (reformat-string "(foo bar\nbaz\nquz)")
           "(foo bar\n     baz\n     quz)"))
```
Proposed new 
```clojure
    (rft "hangs indent to deeper element"
        ["(foo bar"
         "baz"
         "quz)"]
        ["(foo bar"
         "     baz"
         "     quz)"])
```
At a glance, there is no doubt about intended indentation.

## Failures
Suppose I changed `baz` to `biz` in above example for the expected result. Using eftest here's what failures look like.

Current
```
FAIL in cljfmt.core-test/test-indent (core_test.cljc:43)
list indentation
expected: "(foo bar\n     baz\n     quz)"
  actual: ("(foo bar\n     biz\n     quz)")
```
Expected and actual seem reversed to me.

Proposed new 
```
FAIL in cljfmt.core-test/test-indent (core_test.cljc:16)
list indentation
hangs indent to deeper element
expected: ["(foo bar" "     biz" "     quz)"]
  actual: (["(foo bar" "     baz" "     quz)"])
```
Expected and actual are in correct order and test description is shown.

# Comments tests
To get a feel for what these would like, let's dive into a testing block.
Here's what comments tests currently look like:
```clojure
  (testing "comments"
    (is (= (reformat-string ";foo\n(def x 1)")
           ";foo\n(def x 1)"))
    (is (= (reformat-string "(ns foo.core)\n\n;; foo\n(defn foo [x]\n(inc x))")
           "(ns foo.core)\n\n;; foo\n(defn foo [x]\n  (inc x))"))
    (is (= (reformat-string ";; foo\n(ns foo\n(:require bar))")
           ";; foo\n(ns foo\n  (:require bar))"))
    (is (= (reformat-string "(defn foo [x]\n  ;; +1\n(inc x))")
           "(defn foo [x]\n  ;; +1\n  (inc x))"))
    (is (= (reformat-string "(let [;foo\n x (foo bar\n baz)]\n x)")
           "(let [;foo\n      x (foo bar\n             baz)]\n  x)"))
    (is (= (reformat-string "(binding [x 1] ; foo\nx)")
           "(binding [x 1] ; foo\n  x)")))
```
Here's my proposal for change:
```clojure
  (testing "comments"
    (rft "header comment unchanged"
        [ ";foo"
          "(def x 1)"]
        [ ";foo"
         "(def x 1)"])
    (rft "post namespace comment unchanged, code reformatted"
        ["(ns foo.core)"
         ""
         ";; foo"
         "(defn foo [x]"
         "(inc x))"]
        ["(ns foo.core)"
         ""
         ";; foo"
         "(defn foo [x]"
         "  (inc x))"])
    (rft "pre namespace comment unchanged, namespace reformatted"
        [";; foo"
         "(ns foo"
         "(:require bar))"]
        [";; foo"
         "(ns foo"
         "  (:require bar))"])
    (rft "commented out line unchanged, code reformatted"
        ["(defn foo [x]"
         "  ;; +1"
         "(inc x))"]
        ["(defn foo [x]"
         "  ;; +1"
         "  (inc x))"])
    (rft "inline comment unchanged, code reformatted"
        ["(let [;foo"
         " x (foo bar"
         " baz)]"
         " x)"]
        ["(let [;foo"
         "      x (foo bar"
         "             baz)]"
         "  x)"])
    (rft "binding inline comment unchanged, code reformated"
        ["(binding [x 1] ; foo"
         "x)"]
        ["(binding [x 1] ; foo"
         "  x)"]))
```
# Example with options
Here's how I propose a reformat-string test with options would be expressed:
```clojure
    (rft "custom indents applied to namespaced defrecord"
         ["(ns example"
          "(:require [thing.core :as t]))"
          ""
          "(t/defrecord Foo [x]"
          "Closeable"
          "(close [_]"
          "(prn x)))"
          ""
          "(defrecord Foo [x]"
          "Closeable"
          "(close [_]"
          "(prn x)))"]
         {:indents {'thing.core/defrecord [[:inner 0]]}
          #?@(:cljs [:alias-map {"t" "thing.core"}])}
         ["(ns example"
          "    (:require [thing.core :as t]))"
          ""
          "(t/defrecord Foo [x]"
          "  Closeable"
          "  (close [_]"
          "         (prn x)))"
          ""
          "(defrecord Foo [x]"
          "           Closeable"
          "           (close [_]"
          "                  (prn x)))"])
```

### PR DESCRIPTION
**Pull request by lread**
Rework string-format unit tests to be easier to digest by:

 - Giving each input line its own line (no more embbeded \n)

 - Making the intent of each test clear by giving it a description

 - Making failures easier to decipher by switching expected and actual

Fixes #145.

### CODE REVIEW

**Comment by weavejester** [comment]
Thanks for the PR. I seem to recall putting `is` into a different function messes up the error stacktrace or line numbers, so I'd prefer we instead use `assert-expr`. Otherwise everything seems like a good idea. So perhaps instead of:

```clojure
(rft "map"
      ["{:foo 1"
       ":bar 2}"]
      ["{:foo 1"
       " :bar 2}"])
```

We write:

```clojure
(is (reformats-into?
     ["{:foo 1"
      ":bar 2}"]
     ["{:foo 1"
      " :bar 2}"]))
```

Via some `assert-expr` method like:

```clojure
(defmethod assert-expr 'reformats-into? [msg [input-lines expected-lines options]]
  `(let [input#    (str/join "\n" ~input-lines)
         actual#   (reformat-string input# ~(or options {}))
         expected# (str/join "\n" ~expected-lines)
         result#   (= expected# actual#)]
     (if result#
       (do-report {:type :pass, :message ~msg,
                   :expected expected#, :actual actual#})
       (do-report {:type :fail, :message ~msg,
                   :expected expected#, :actual actual#}))
     result#)))
```

**Comment by lread** [comment]
Thanks for your thoughtful and kind review!  In addition to dealing with error stack trace line number issues, I think your test syntax is is more expressive than my original shot at it. I will take a look at making this change soon.

**Comment by lread** [comment]
My macro fu leveled up slightly. [Portable Custom clojure.test Assertions](http://seespotcode.net/2018/01/13/portable-clojure-test-assert-expr/) helped me quite a bit!

For your consideration:
1. I kept my comparison on vectors of lines because I preferred its output on error over a single string (I assume you might have been suggesting single string from your `assert-expr` sample in above comment.) Here's what an error looks like using eftest: 
```
FAIL in cljfmt.core-test/test-indent (core_test.cljc:233)
namespaces
applies custom indentation to namespaced defrecord
expected: ["(ns example"
           "    (:require [thing.core :as t]))"
           ""
           "(t/defrecord Boo [x]"
           "  Closeable"
           "  (close [_]"
           "         (prn x)))"
           ""
           "(defrecord Foo [x]"
           "           Closeable"
           "           (close [_]"
           "                  (prn x)))"]
  actual: ["(ns example"
           "    (:require [thing.core :as t]))"
           ""
           "(t/defrecord Foo [x]"
           "  Closeable"
           "  (close [_]"
           "         (prn x)))"
           ""
           "(defrecord Foo [x]"
           "           Closeable"
           "           (close [_]"
           "                  (prn x)))"]
```
2. I chose the shorter `reformats-to?` over suggested `reformats-into?`.
3. My original commit forced a description on each test. While reworking to assert expressions, I realized that this was overkill and only provided descriptions where I felt they would genuinely help.

**Comment by lread** [comment]
After we are all good and this is merged I was thinking of issues #82, #111 and #112.  I have taken a peek and it seems to me they have been almost (but not entirely) fixed by other changes. I'll add proper unit test coverage.

File: cljfmt/test/cljfmt/core_test.cljc [hunk 1]
```diff
@@ -4,289 +4,744 @@
...
+      "discard")
+  (is (reformats-to?
+       ["(juxt +' -')"]
+       ["(juxt +' -')"])
+      "arbirtrary sum of sums and negation")
+  (is (reformats-to?
+       ["#\"(?i)foo\""]
+       ["#\"(?i)foo\""])
+      "regular expression")
+  ; newline is embedded in regex, this test is more readable if we don't use our assert wrapper
```
Comment (weavejester) [inline cljfmt/test/cljfmt/core_test.cljc]: This comment can be removed, as I don't think it's necessary to understand the code.

File: cljfmt/test/cljfmt/assert/cljs.cljc [hunk 2]
```diff
@@ -0,0 +1,10 @@
+(ns cljfmt.assert.cljs
+  (:require [cljs.test :as test]
+            [cljfmt.assert.impl :as impl]))
+
+;; cljs macros are compiled by clj
+#?(:clj
```
Comment (weavejester) [inline cljfmt/test/cljfmt/assert/cljs.cljc]: Why do you have `cljfmt.assert.cljs`, `cljfmt.assert.impl` and `cljfmt.asserts`? Can you put everything into `cljfmt.core-test`?

Comment (lread) [reply cljfmt/test/cljfmt/assert/cljs.cljc]: My code was heavily inspired by [this example](https://github.com/grzm/ex.assertions) which splits things out in a very similar way. I personally prefer to move self contained complexity out to separate files.  I find when source files get larger they also get harder to digest and maintain. I realize this is a subjective call and am happy to move to your preference.

`asserts.cljc` could definitely be merged into `core_test.cljc`, but I'm thinking there might be some file separation gymnastics needed to support test assertions under ClojureScript: `assert/cljs.cljc` is required for ClojureScript but is compiled by Clojure. I am new to this area though, I'll try to understand what is possible and get back to you.

You did not ask but: I added the `assert` sub-directory to try to convey these are internal details that `core_test` need not concern itself with directly.

Comment (weavejester) [reply cljfmt/test/cljfmt/assert/cljs.cljc]: I'd rather just move the method into `core_test.cljc`, or possibly `test_util.cljc`. The way the cljs macros interact might make placing it in a secondary namespace a good idea. However, the namespace should indicate clearly that it's a test namespace, hence `test-util`. You should also be able to do it with only one `.cljc` file.

Comment (lread) [reply cljfmt/test/cljfmt/assert/cljs.cljc]: I have not succeeding at merging everything into `core_test.cljc`. I did manage to merge `asserts.cljc` into `core_test.cljc`.  I'm left with the cljs `assert-expr` in a separate source file, and the code that the clj and cljs `assert-expr`s share in another file.  My last push called these `assert_cljs.cljc` and `assert_impl.cljc` respectively.  Do you think moving these two files under a `test-util` subdir would work for you?

BTW, thanks for your patience and encouragement, it is much appreciated.

Comment (lread) [reply cljfmt/test/cljfmt/assert/cljs.cljc]: Thanks for clearing that up, I was thinking I was missing some concept or other.  

If I understand correctly, which I might not, `common` and `cljs` files need the `cljc` extension. 

Your new separation makes usage from ns in `core-tests` very clear.  I also like your tweaks to the code, you have a real talent for succinctness.

Note: I've left the order of args when including options as: `in-line options expected-lines`. I feel they belong next to `in-lines` for better readability of tests.

Comment (weavejester) [reply cljfmt/test/cljfmt/assert/cljs.cljc]: > If I understand correctly, which I might not, `common` and `cljs` files need the `cljc` extension.

What's your reasoning? If `assert-expr` is used in `is`, and `is` is a macro, doesn't that mean that only Clojure files are necessary?

Comment (lread) [reply cljfmt/test/cljfmt/assert/cljs.cljc]: I misled you, I don't understand yet. Let me rephrase: I do not know why, but to get the build to succeed, I needed to rename `common` and `cljs` files with the `cljc` extension.

Comment (weavejester) [reply cljfmt/test/cljfmt/assert/cljs.cljc]: Okay, lets set them as `cljc` then.

Comment (lread) [reply cljfmt/test/cljfmt/assert/cljs.cljc]: Ok, I've pushed changes, they are ready for your review.

File: cljfmt/test/cljfmt/test_util/common.cljc [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+(defn assert-reformats-to
+  ([msg in-lines expected-lines]
+   (assert-reformats-to msg in-lines {} expected-lines))
+  ([msg in-lines options expected-lines]
+   (let [input (str/join "\n" in-lines)
+         actual (reformat-string input options)
+         preserve-trailing-empty-strings -1
+         actual-lines (str/split actual #"\n" preserve-trailing-empty-strings)]
+     {:type (if (= expected-lines actual-lines) :pass :fail)
+      :message msg,
```
Comment (weavejester) [inline cljfmt/test/cljfmt/test_util/common.cljc]: Can you remove this extra `,`

File: cljfmt/test/cljfmt/test_util/common.cljc [hunk 4]
```diff
@@ -0,0 +1,16 @@
+(ns cljfmt.test-util.common
+  (:require [clojure.string :as str]
+            [cljfmt.core :refer [reformat-string]]))
+
+(defn assert-reformats-to
+  ([msg in-lines expected-lines]
+   (assert-reformats-to msg in-lines {} expected-lines))
+  ([msg in-lines options expected-lines]
```
Comment (weavejester) [inline cljfmt/test/cljfmt/test_util/common.cljc]: I think I'd prefer to have the options last. I understand why you put the `expected-lines` there, but I think the argument list is more predictable if the positioning doesn't change.

File: cljfmt/test/cljfmt/test_util/common.cljc [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+   (let [input (str/join "\n" in-lines)
+         actual (reformat-string input options)
+         preserve-trailing-empty-strings -1
```
Comment (weavejester) [inline cljfmt/test/cljfmt/test_util/common.cljc]: I don't think we necessarily need this, since someone can just look up what `-1` means in this context.

File: cljfmt/test/cljfmt/test_util/common.cljc [hunk 6]
```diff
@@ -0,0 +1,16 @@
+(ns cljfmt.test-util.common
+  (:require [clojure.string :as str]
+            [cljfmt.core :refer [reformat-string]]))
+
+(defn assert-reformats-to
+  ([msg in-lines expected-lines]
+   (assert-reformats-to msg in-lines {} expected-lines))
+  ([msg in-lines options expected-lines]
+   (let [input (str/join "\n" in-lines)
```
Comment (weavejester) [inline cljfmt/test/cljfmt/test_util/common.cljc]: Can you ensure the `let` fields are aligned?

File: cljfmt/test/cljfmt/test_util/common.cljc [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+         actual (reformat-string input options)
+         preserve-trailing-empty-strings -1
+         actual-lines (str/split actual #"\n" preserve-trailing-empty-strings)]
+     {:type (if (= expected-lines actual-lines) :pass :fail)
```
Comment (weavejester) [inline cljfmt/test/cljfmt/test_util/common.cljc]: Can you ensure the map fields are aligned?

File: cljfmt/test/cljfmt/core_test.cljc [hunk 8]
```diff
@@ -4,289 +4,750 @@
             [cljfmt.core :refer [reformat-string default-line-separator
                                  normalize-newlines find-line-separator
                                  replace-newlines wrap-normalize-newlines]]
-            [clojure.string :as str]))
+            [cljfmt.test-util.common :as common])
+  #?(:clj (:require [cljfmt.test-util.clojure])
+     :cljs (:require-macros [cljfmt.test-util.cljs])))
+
+;;; `reformats-to?` makes `reformat-string` tests more readable.
```
Comment (weavejester) [inline cljfmt/test/cljfmt/core_test.cljc]: I think these lines can be removed, as the tests themselves are good examples of use.

File: cljfmt/test/cljfmt/core_test.cljc [hunk 9]
```diff
@@ -4,289 +4,750 @@
             [cljfmt.core :refer [reformat-string default-line-separator
                                  normalize-newlines find-line-separator
                                  replace-newlines wrap-normalize-newlines]]
-            [clojure.string :as str]))
+            [cljfmt.test-util.common :as common])
+  #?(:clj (:require [cljfmt.test-util.clojure])
```
Comment (weavejester) [inline cljfmt/test/cljfmt/core_test.cljc]: Rather than having two `:require` statements, it would be better just to have one.

File: cljfmt/test/cljfmt/core_test.cljc [hunk 10]
```diff
@@ -4,289 +4,750 @@
...
-  (is (= (reformat-string "(juxt +' -')") "(juxt +' -')"))
-  (is (= (reformat-string "#\"(?i)foo\"") "#\"(?i)foo\""))
-  (is (= (reformat-string "#\"a\nb\"") "#\"a\nb\"")))
+  (is (reformats-to?
+       [";foo"]
+       [";foo"]))
+  (is (reformats-to?
+       ["::foo"]
+       ["::foo"])
+      "auto-resolved keyword")
```
Comment (weavejester) [inline cljfmt/test/cljfmt/core_test.cljc]: I don't think we need messages where the reformatting is obvious or trivial.

File: cljfmt/test/cljfmt/test_util/common.cljc (hunk 4 above)
Comment (lread) [reply cljfmt/test/cljfmt/test_util/common.cljc]: I see your point and realize I don't feel too strongly about my preference.

File: cljfmt/test/cljfmt/test_util/common.cljc (hunk 5 above)
Comment (lread) [reply cljfmt/test/cljfmt/test_util/common.cljc]: I didn't see it clearly documented, but now see it in user examples. Will comply.

File: cljfmt/test/cljfmt/core_test.cljc (hunk 10 above)
Comment (lread) [reply cljfmt/test/cljfmt/core_test.cljc]: I am certainly a newbie in some areas of Clojure, if I looked it up to double check what it is, I probably added a msg.  I'll review.

**Comment by weavejester** [comment]
Thanks, this looks all good. Can you squash your commits down? Then I should be able to merge.

**Comment by lread** [comment]
squash in progress...

**Comment by lread** [comment]
Done. Maybe. Please let me know if I've squashed to expectations.  Thanks.

**Comment by weavejester** [comment]
I'm not quite sure what you've done, but you've changed several existing commits and added in a superfluous merge commit. Try running the following commands:

    git reset --hard origin/master
    git cherry-pick --strategy=recursive -Xtheirs 28a13c8d4bb8215dba6b3c40e575cda9a16cb82f
    git push -f

**Comment by lread** [comment]
Oh sorry about that, my rebasing is obviosuly very very rusty.

**Comment by lread** [comment]
Thanks, you saved me!

**Comment by weavejester** [comment]
No problem. I think there's just one more thing to do, which is to adjust the commit message slightly. You can change it with:

    git commit --amend

Change it to:

```
Refactor reformat tests to be more readable

Adds a 'reformats-to?' custom assertion. Fixes #145.
```

I think everything else is evident from the diff.

**Comment by lread** [comment]
Ok by me. Done.
